src/model_theory/terms_and_formulas.lean

```diff
@@ -128,36 +129,76 @@ variables {L} {α : Type}
 section formula
 variable {n : ℕ}
 
-@[simps] instance : has_bot (L.bounded_formula α n) := ⟨bd_falsum⟩
+namespace bounded_formula
+
+instance : has_bot (L.bounded_formula α n) := ⟨bd_falsum⟩
 
 /-- The negation of a bounded formula is also a bounded formula. -/
-@[reducible] def bd_not (φ : L.bounded_formula α n) : L.bounded_formula α n := bd_imp φ ⊥
+def bd_not (φ : L.bounded_formula α n) : L.bounded_formula α n := bd_imp φ ⊥
+
+/-- Puts an `∃` quantifier on a bounded formula. -/
+def bd_exists (φ : L.bounded_formula α (n + 1)) : L.bounded_formula α n :=
+  bd_not (bd_all (bd_not φ))
+
+instance : has_top (L.bounded_formula α n) := ⟨bd_not ⊥⟩
 
-@[simps] instance : has_top (L.bounded_formula α n) := ⟨bd_not bd_falsum⟩
+instance : has_inf (L.bounded_formula α n) := ⟨λ f g, bd_not (bd_imp f (bd_not g))⟩
 
-@[simps] instance : has_inf (L.bounded_formula α n) := ⟨λ f g, bd_not (bd_imp f (bd_not g))⟩
+instance : has_sup (L.bounded_formula α n) := ⟨λ f g, bd_imp (bd_not f) g⟩
 
-@[simps] instance : has_sup (L.bounded_formula α n) := ⟨λ f g, bd_imp (bd_not f) g⟩
+/-- The biimplication between two bounded formulas. -/
+def bd_iff (φ ψ : L.bounded_formula α n) := φ.bd_imp ψ ⊓ ψ.bd_imp φ
+
+/-- A function to help relabel the variables in bounded formulas. -/
+def relabel_aux {n : ℕ} {α β : Type} (g : α → (β ⊕ fin n)) (k : ℕ) :
+  α ⊕ fin k → β ⊕ fin (n + k) :=
+(sum.map id fin_sum_fin_equiv) ∘ (equiv.sum_assoc _ _ _) ∘ (sum.map g id)
```

Comment:
this is a very unfortunate consequence of the design decisions in `bounded_formula`... Not sure if this can be done in a nicer way.

Reply:
I did this to generalize two operations that were separately much nicer-looking - turning a `formula` back into a `bounded_formula`, thus making some of its variables eligible to be quantified, and the relabeling of free variables that I do for `formula`s. Perhaps in this case, it is better to just do them separately.

Reply:
On a further attempt, turning a formula back into a bounded formula really is almost this bad.